File: leetcode/easy/problem_94.py
"""
94. Binary Tree Inorder Traversal

Given the root of a binary tree, return the inorder traversal of its nodes' values.
"""
from typing import Optional, List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution1:

    results = list()

    def rec(self, root):
        if root:
            self.rec(root.left)
            self.results.append(root.val)
            self.rec(root.right)

    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        self.rec(root)
        return self.results


class Solution2:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:

        if root is None:
            return list()

        stack = [[root, True]]
        results = list()

        while True:
            if len(stack) == 0:
                break
            root, go_left = stack[-1]

            # if we still have left and if we still
            # didn't go there, do so
            if root.left is not None and go_left:
                stack.append([root.left, True])
            else:
                # if there is no more left, pop the last node
                # and append it to results
                results.append(stack.pop()[0].val)
                # else if there is no more left and
                # if we have something in the stack,
                # turn the go_left flag into True
                if len(stack) != 0:
                    stack[-1][-1] = False
                # if there is right from the last root, go there
                if root.right is not None:
                    stack.append([root.right, True])
        return results

File: leetcode/easy/problem_94.cpp
/*
94. Binary Tree Inorder Traversal

Given the root of a binary tree, return the inorder traversal of its nodes' values.

Example 1:

Input: root = [1,null,2,3]
Output: [1,3,2]

Example 2:

Input: root = []
Output: []

Example 3:

Input: root = [1]
Output: [1]
*/

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    void inorderHelper(TreeNode* root, vector<int>& result) {
        if (root == nullptr) {
            return;
        }
        inorderHelper(root->left, result);
        result.push_back(root->val);
        inorderHelper(root->right, result);
    }

    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> result;
        inorderHelper(root, result);
        return result;
    }
};